# Iomega workflow
## Using matchms to harmonize, clean, and complete metadata
Here we run matchms filters and additional custom filters to do extensive cleaning of a large spectral dataset (>140,000 spectra).

In [1]:
import os
import sys

ROOT = os.path.dirname(os.getcwd())
path_data = os.path.join(ROOT, 'data')
sys.path.insert(1, ROOT)

### Import data from minimally processed mgf

In [2]:
from matchms.importing import load_from_mgf

mgf_file = os.path.join(path_data, 'all_gnps_added_metadata_fixed.mgf')

reference_spectrums_raw = load_from_mgf(mgf_file)

RDKit WARNING: [09:55:09] Enabling RDKit 2019.09.3 jupyter extensions


### Run basic filters on all spectra

In [3]:
from matchms.filtering import default_filters
from matchms.filtering import add_parent_mass

def apply_filters(s):
    s = default_filters(s)
    s = add_parent_mass(s)
    return s

reference_spectrums = [apply_filters(s) for s in reference_spectrums_raw]

Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+

Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+2H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+H
Added i

Added ionmode 'negative' based on adduct:  M-H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'negative' based on adduct:  M-H
Added ionmode 'negative' based on adduct:  M-H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+H
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
A

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3C

Added ionmode 'positive' based on adduct:  M+NH4+
Added ionmode 'positive' based on adduct:  M+NH4+
Added ionmode 'positive' based on adduct:  M+NH4+
Added ionmode 'positive' based on adduct:  M+NH4+
Added ionmode 'positive' based on adduct:  M+NH4+
Added ionmode 'positive' based on adduct:  M+NH4+
Added ionmode 'positive' based on adduct:  M+NH4+
Added ionmode 'positive' based on adduct:  M+NH4+
Added ionmode 'positive' based on adduct:  M+NH4+
Added ionmode 'positive' based on adduct:  M+NH4+
Added ionmode 'positive' based on adduct:  M+NH4+
Added ionmode 'positive' based on adduct:  M+NH4+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'n

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-/M-Ser-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-/M-Ser-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on

Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M-H-/M-Ser-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-/M-Ser-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3C

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-/M-Ser-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M-H-/M-Ser-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  (M+CH3COOH)-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  (M+CH3COOH)-H-
Added ionmode 'negative' based on adduct:  (M+CH3COOH)-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionm

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'n

Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH

Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-/M-Ser-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-/M-Ser-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on addu

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M-H-/M-Ser-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added io

Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M-H-
Added ionm

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M+CH3COO-/M-CH3-
Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct

Added ionmode 'negative' based on adduct:  M+CH3COO-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on a

Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'negative' based on adduct:  M-H-
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct:  M+H+
Added ionmode 'positive' based on adduct

## Clean (and where possible extend) metadata

### 1) Harmization
+ Here, undefiend entries will be harmonized (instead of having a huge variation of None,"", "N/A" etc.)
+ The ``repair_inchi_inchikey_smiles`` function will correct misplaced metadata (e.g. inchikeys entered as inchi etc.) and harmonize the entry strings.

In [4]:
from matchms.filtering import harmonize_undefined_inchikey, harmonize_undefined_inchi, harmonize_undefined_smiles
from matchms.filtering import repair_inchi_inchikey_smiles

def clean_metadata(s):
    s = harmonize_undefined_inchikey(s)
    s = harmonize_undefined_inchi(s)
    s = harmonize_undefined_smiles(s)
    s = repair_inchi_inchikey_smiles(s)
    return s

reference_spectrums = [clean_metadata(s) for s in reference_spectrums]

### Check status of metadata annotations
Note: Unique InchiKey here means: unique in the first 14 characters.

In [5]:
def count_annotations(spectra):
    inchi_lst = []
    smiles_lst = []
    inchikey_lst = []
    for i, spec in enumerate(spectra):
        inchi_lst.append(spec.get("inchi"))
        smiles_lst.append(spec.get("smiles"))
        inchikey_lst.append(spec.get("inchikey"))

    inchi_count = sum([1 for x in inchi_lst if x])
    smiles_count = sum([1 for x in smiles_lst if x])
    inchikey_count = sum([1 for x in inchikey_lst if x])
    print("Inchis:", inchi_count, "--", len(set(inchi_lst)), "unique")
    print("Smiles:", smiles_count, "--", len(set(smiles_lst)), "unique")
    print("Inchikeys:", inchikey_count, "--", 
          len(set([x[:14] for x in inchikey_lst if x])), "unique")

count_annotations(reference_spectrums)

Inchis: 77147 -- 10267 unique
Smiles: 38221 -- 14509 unique
Inchikeys: 77343 -- 10035 unique


### 2) Convert entries where possible
Where possible (and necessary, i.e. missing): Convert between smiles, inchi, inchikey to complete metadata. This is done using functions from rdkit.

In [6]:
from matchms.filtering import derive_inchi_from_smiles, derive_smiles_from_inchi
from matchms.filtering import derive_inchikey_from_inchi

def clean_metadata2(s):
    s = derive_inchi_from_smiles(s)
    s = derive_smiles_from_inchi(s)
    s = derive_inchikey_from_inchi(s)
    return s

reference_spectrums = [clean_metadata2(s) for s in reference_spectrums]

### Check improvement of metadata:

In [7]:
count_annotations(reference_spectrums)

Inchis: 87739 -- 15005 unique
Smiles: 87195 -- 18943 unique
Inchikeys: 87734 -- 12939 unique


In [8]:
from matchms.exporting.save_as_json import save_as_json

save_as_json(reference_spectrums, os.path.join(path_data,'gnps_all_after_matchms_cleaning.json'))

In [2]:
from matchms.importing.load_from_json import load_from_json
reference_spectrums = load_from_json(os.path.join(path_data,'gnps_all_after_matchms_cleaning.json'))

RDKit WARNING: [21:55:32] Enabling RDKit 2019.09.3 jupyter extensions


### 3) Complete metadata using pubchem lookup routines
Try to complete and extend metadata based on looking for matches in PubChem. Found matches on PubChem will be accepted if:
+ **Single matching extry** for InchiKey (first 14 characters)
+ **Single matching extry** for Inchi
+ **Single matching extry** for Smiles
+ Matching name AND inchikey
+ Matching name AND inchi
+ Matching molecular formula AND inchi or inchikey
+ **Single matching entry** with same name AND molecular weight == parent mass.

### Carefull: when comparing InchiKeys we here look at the first 14 characters only. So stereo-chemistry is ignored.

## Here I only ran the first 20,000 spectra since this part is SLOW

In [7]:
from custom_functions.pubchem_lookup import lookup_metadata_completion

reference_spectrums_lookup = []
for i, s in enumerate(reference_spectrums[:20000]):
    if i/1000 == i//1000:  # show progress, can be slow...
        print("spectrum no.", i, 20*"--")
    reference_spectrums_lookup.append(lookup_metadata_completion(s, search_depth=10))

spectrum no. 0 ----------------------------------------
spectrum no. 1000 ----------------------------------------
Added metadata based on: 
 Matching inchi (>= 3 parts).Matching inchikey (>= 1 parts).
spectrum no. 2000 ----------------------------------------
Added metadata based on: 
 Matching molecular weight (544.6 vs parent mass 544.0).Matching compound name (Scytonemin).
Added metadata based on: 
 Matching molecular weight (729.9 vs parent mass 729.4).Matching compound name (Exumolide A).
Added metadata based on: 
 Matching molecular weight (512.6 vs parent mass 512.3).Matching compound name (Microsporin A).
Added metadata based on: 
 Matching molecular weight (483.6 vs parent mass 483.0).Matching compound name (Tumonoic Acid H).
Added metadata based on: 
 Matching molecular weight (671.8 vs parent mass 671.1).Matching compound name (Palmyramide A).
Added metadata based on: 
 Matching molecular weight (497.7 vs parent mass 496.9).Matching compound name (Tumonoic Acid I).
Added me

Added metadata based on: 
 Matching molecular weight (244.3 vs parent mass 244.1).Matching compound name (cyclo(L-Phe-L-Pro)).
Added metadata based on: 
 Matching molecular weight (210.3 vs parent mass 210.1).Matching compound name (cyclo(L-Leu-L-Pro)).
Added metadata based on: 
 Matching molecular weight (210.3 vs parent mass 210.1).Matching compound name (cyclo(L-Leu-L-Pro)).
Added metadata based on: 
 Matching molecular weight (210.3 vs parent mass 210.1).Matching compound name (cyclo(L-Leu-L-Pro)).
Added metadata based on: 
 Matching molecular weight (210.3 vs parent mass 210.1).Matching compound name (cyclo(L-Leu-L-Pro)).
Added metadata based on: 
 Matching molecular weight (210.3 vs parent mass 210.1).Matching compound name (cyclo(L-Leu-L-Pro)).
Added metadata based on: 
 Matching molecular weight (555.8 vs parent mass 555.3).Matching compound name (pateamine).
Added metadata based on: 
 Matching molecular weight (555.8 vs parent mass 555.3).Matching compound name (pateamine).
Ad

Added metadata based on: 
 Matching molecular weight (254.4 vs parent mass 254.2).Matching compound name (Hexadecenoic acid).
Added metadata based on: 
 Matching molecular weight (296.5 vs parent mass 296.3).Matching compound name (Lactobacillic acid).
Added metadata based on: 
 Matching molecular weight (408.6 vs parent mass 408.3).Matching compound name (Cholic Acid).
Added metadata based on: 
 Matching molecular weight (494.6 vs parent mass 494.4).Matching compound name (Alteramide B).
Added metadata based on: 
 Matching molecular weight (391.4 vs parent mass 391.6).Matching compound name (Estatin A).
Added metadata based on: 
 Matching molecular weight (407.4 vs parent mass 407.5).Matching compound name (Estatin B).
Added metadata based on: 
 Matching molecular weight (624.8 vs parent mass 624.4).Matching compound name (Citrinadin A).
Added metadata based on: 
 Matching molecular weight (610.8 vs parent mass 610.4).Matching compound name (Citrinadin C).
Added metadata based on: 
 M

Added metadata based on: 
 Matching molecular weight (1092.4 vs parent mass 1091.7).Matching compound name (Barangamide A).
Added metadata based on: 
 Matching molecular weight (1376.8 vs parent mass 1375.9).Matching compound name (Theonellapeptolide V).
Added metadata based on: 
 Matching molecular weight (1092.4 vs parent mass 1091.7).Matching compound name (Barangamide A).
Added metadata based on: 
 Matching molecular weight (1390.8 vs parent mass 1389.9).Matching compound name (Theonellapeptolide IId).
Added metadata based on: 
 Unique Smiles match (N/C3=[NH+]/C(=C2\CCNC(=O)c1nc(Br)c(Br)c12)C(=O)N3).
Added metadata based on: 
 Unique Smiles match (N/C3=[NH+]/C(=C2\CCNC(=O)c1nc(Br)c(Br)c12)C(=O)N3).
Added metadata based on: 
 Matching molecular weight (764.0 vs parent mass 763.5).Matching compound name (PE(16:0/22:6)).
Added metadata based on: 
 Matching molecular weight (790.1 vs parent mass 789.5).Matching compound name (PE(18:1/22:6)).
Found no matches based on smiles.
Found no m

Added metadata based on: 
 Unique Smiles match (OCC\C(=C\C1=[O][Fe]2345[O]=C(\C=C(/C)CCO)N(CCCC6NC(=O)CNC(=O)C(NC(=O)C(NC(C(CCCN1O2)NC(C(CCCN(C(\C=C(/C)CCO)=[O]3)O4)NC6=O)=O)=O)CO)CO)O5)C).
Added metadata based on: 
 Unique Smiles match (CC1=[O][Fe]2345ON1CCCC1NC(=O)CNC(=O)C(CO)NC(=O)C(CO)NC(=O)C(CCCN(O2)C(C)=[O]3)NC(=O)C(CCCN(O4)C(C)=[O]5)NC1=O).
Added metadata based on: 
 Unique Smiles match (C\C1=C\C2=[O][Fe]3456ON2CCCC(N)C(=O)OCC\C(C)=C/C(=[O]3)N(CCCC(N)C(=O)OCC\C(C)=C/C(=[O]4)N(CCCC(N)C(=O)OCC1)O5)O6).
Added metadata based on: 
 Unique Smiles match (C\C(CCO)=C\C1=[O][Fe]2345ON(CCCC(N)C(O)=O)C(\C=C(C)/CCOC(=O)C(N)CCCN(O2)C(\C=C(C)/CCOC(=O)C(N)CCCN1O3)=[O]4)=[O]5).
Added metadata based on: 
 Unique Smiles match (CC(=O)NC1CCCN2O[Fe]345(ON(CCCC6NC(=O)C(CCCN(O3)C(\C=C(C)\CCOC1=O)=[O]4)NC6=O)C(\C=C(/C)CCO)=[O]5)[O]=C2\C=C(/C)CCO).
Added metadata based on: 
 Unique Smiles match (CC(=O)NC1CCCN2O[Fe]345(ON(CCCC6NC(=O)C(CCCN(O3)C(\C=C(C)/CCOC1=O)=[O]4)NC6=O)C(C)=[O]5)[O]=C2\C=C(/C)CCO).
Add

Added metadata based on: 
 Matching molecular weight (470.5 vs parent mass 470.2).Matching compound name (Physodic acid).
Added metadata based on: 
 Matching molecular weight (292.3 vs parent mass 292.1).Matching compound name (Polyporic acid).
Added metadata based on: 
 Matching molecular weight (416.5 vs parent mass 416.2).Matching compound name (Sphaerophorin).
Added metadata based on: 
 Matching molecular weight (344.3 vs parent mass 344.1).Matching compound name (Usnic acid).
Added metadata based on: 
 Matching molecular weight (324.4 vs parent mass 324.1).Matching compound name (Pyochelin).
Added metadata based on: 
 Matching molecular weight (324.4 vs parent mass 324.0).Matching compound name (Pyochelin).
Added metadata based on: 
 Matching molecular weight (193.2 vs parent mass 193.0).Matching compound name (Toxoflavin).
Added metadata based on: 
 Matching molecular weight (210.2 vs parent mass 210.1).Matching compound name (Pyocyanin).
Added metadata based on: 
 Matching molec

Added metadata based on: 
 Matching molecular weight (333.3 vs parent mass 333.1).Matching compound name ( Trp-Glu).
Added metadata based on: 
 Matching molecular weight (161.2 vs parent mass 161.1).Matching compound name ( L-Carnitine).
Added metadata based on: 
 Matching molecular weight (307.3 vs parent mass 307.1).Matching compound name ( L-Glutathione, reduced).
Added metadata based on: 
 Matching molecular weight (188.2 vs parent mass 188.1).Matching compound name ( Gly-Leu).
Added metadata based on: 
 Matching molecular weight (181.2 vs parent mass 181.1).Matching compound name ( L-Tyrosine).
Added metadata based on: 
 Matching molecular weight (362.5 vs parent mass 362.2).Matching compound name ( Hydrocortisone).
Added metadata based on: 
 Matching molecular weight (286.2 vs parent mass 286.0).Matching compound name ( Datiscetin).
Added metadata based on: 
 Matching molecular weight (333.3 vs parent mass 333.1).Matching compound name ( Trp-Glu).
Added metadata based on: 
 Match

Added metadata based on: 
 Matching molecular weight (310.5 vs parent mass 310.3).Matching compound name ( cis-11-Eicosenoic acid).
Added metadata based on: 
 Matching molecular weight (244.3 vs parent mass 244.2).Matching compound name ( Leu-Leu).
Added metadata based on: 
 Matching molecular weight (723.0 vs parent mass 722.5).Matching compound name ( 1,2-Dipalmitoyl-sn-glycero-3-phospho-(1'-rac-glycerol)).
Added metadata based on: 
 Matching molecular weight (318.3 vs parent mass 318.1).Matching compound name ( Phenolphthalein).
Added metadata based on: 
 Matching molecular weight (502.6 vs parent mass 502.3).Matching compound name ( Undecaethylene glycol).
Added metadata based on: 
 Matching molecular weight (284.3 vs parent mass 284.1).Matching compound name ( Biochanin A).
Added metadata based on: 
 Matching molecular weight (354.3 vs parent mass 354.1).Matching compound name ( Chlorogenic acid).
Added metadata based on: 
 Matching molecular weight (310.3 vs parent mass 310.1).Ma

Added metadata based on: 
 Matching molecular weight (290.3 vs parent mass 290.1).Matching compound name ( Trimethoprim).
Added metadata based on: 
 Matching molecular weight (740.0 vs parent mass 739.5).Matching compound name ( 1,2-Dilinoleoyl-sn-glycero-3-phosphoethanolamine).
Added metadata based on: 
 Matching molecular weight (373.4 vs parent mass 373.2).Matching compound name ( Spirotetramat).
Added metadata based on: 
 Matching molecular weight (318.4 vs parent mass 318.2).Matching compound name ( 12-Oxo-5Z,8Z,10E,14Z-eicosatetraenoic acid).
Added metadata based on: 
 Matching molecular weight (549.7 vs parent mass 549.4).Matching compound name ( 1-O-Hexadecyl-2-O-(2E-butenoyl)-sn-glyceryl-3-phosphocholine).
Added metadata based on: 
 Matching molecular weight (278.4 vs parent mass 278.2).Matching compound name ( Pinolenic acid).
Added metadata based on: 
 Matching molecular weight (509.7 vs parent mass 509.3).Matching compound name ( 1-Heptadecanoyl-sn-glycero-3-phosphocholine)

Added metadata based on: 
 Matching molecular weight (192.2 vs parent mass 192.1).Matching compound name ( (-)Quinic acid).
Added metadata based on: 
 Matching molecular weight (665.4 vs parent mass 665.1).Matching compound name ( .beta.-Nicotinamide adenine dinucleotide, reduced).
Added metadata based on: 
 Matching molecular weight (509.7 vs parent mass 509.4).Matching compound name ( Lyso-PAF C-18).
Added metadata based on: 
 Matching molecular weight (294.4 vs parent mass 294.2).Matching compound name ( 13S-Hydroxy-9Z,11E,15Z-octadecatrienoic acid).
Added metadata based on: 
 Matching molecular weight (259.4 vs parent mass 259.2).Matching compound name ( Lys-Ile).
Added metadata based on: 
 Matching molecular weight (290.3 vs parent mass 290.1).Matching compound name ( Trimethoprim).
Added metadata based on: 
 Matching molecular weight (453.5 vs parent mass 453.3).Matching compound name ( 1-Palmitoyl-2-hydroxy-sn-glycero-3-phosphoethanolamine).
Added metadata based on: 
 Matching m

Added metadata based on: 
 Matching molecular weight (155.2 vs parent mass 155.1).Matching compound name ( L-Histidine).
Added metadata based on: 
 Matching molecular weight (258.3 vs parent mass 258.1).Matching compound name ( Rhapontigenin).
Added metadata based on: 
 Matching molecular weight (194.2 vs parent mass 194.1).Matching compound name ( 3-Hydroxy-4-methoxycinnamic acid).
Added metadata based on: 
 Matching molecular weight (390.6 vs parent mass 390.3).Matching compound name ( Dioctyl phthalate).
Added metadata based on: 
 Matching molecular weight (792.1 vs parent mass 791.5).Matching compound name ( 1-Octadecanoyl-2-(4Z,7Z,10Z,13Z,16Z,19Z-docosahexaenoyl)-sn-glycero-3-phosphoethanolamine).
Added metadata based on: 
 Matching molecular weight (218.2 vs parent mass 218.1).Matching compound name ( Ser-Ile).
Added metadata based on: 
 Matching molecular weight (307.3 vs parent mass 307.1).Matching compound name ( L-Glutathione, reduced).
Added metadata based on: 
 Matching mol

Added metadata based on: 
 Matching molecular weight (663.4 vs parent mass 663.1).Matching compound name ( .beta.-Nicotinamide adenine dinucleotide).
Added metadata based on: 
 Matching molecular weight (270.3 vs parent mass 270.1).Matching compound name ( 5-Hydroxy-7-methoxyflavanone).
Added metadata based on: 
 Matching molecular weight (750.1 vs parent mass 749.6).Matching compound name ( 1-Palmitoyl-2-thiopalmitoyl phosphatidylcholine).
Added metadata based on: 
 Matching molecular weight (229.3 vs parent mass 229.0).Matching compound name ( Dimethoate).
Added metadata based on: 
 Matching molecular weight (796.1 vs parent mass 795.6).Matching compound name ( 1-Heptadecanoyl-2-(5Z,8Z,11Z,14Z-eicosatetraenoyl)-sn-glycero-3-phosphocholine).
Added metadata based on: 
 Matching molecular weight (430.7 vs parent mass 430.4).Matching compound name ( (+)-.alpha.-Tocopherol).
Added metadata based on: 
 Matching molecular weight (228.3 vs parent mass 228.1).Matching compound name ( Pro-Ile)

Added metadata based on: 
 Matching molecular weight (749.0 vs parent mass 748.5).Matching compound name ( 1-Hexadecanoyl-2-(9Z-octadecenoyl)-sn-glycero-3-phospho-(1'-rac-glycerol)).
Added metadata based on: 
 Matching molecular weight (294.3 vs parent mass 294.1).Matching compound name ( Phe-Glu).
Added metadata based on: 
 Matching molecular weight (268.3 vs parent mass 268.1).Matching compound name ( 4'-Hydroxy-5-methoxyflavone).
Added metadata based on: 
 Matching molecular weight (166.2 vs parent mass 166.1).Matching compound name ( Cyromazine).
Added metadata based on: 
 Matching molecular weight (205.2 vs parent mass 205.1).Matching compound name ( Dexpanthenol).
Added metadata based on: 
 Matching molecular weight (351.4 vs parent mass 351.0).Matching compound name ( Meloxicam).
Added metadata based on: 
 Matching molecular weight (267.3 vs parent mass 267.1).Matching compound name ( Diethofencarb).
Added metadata based on: 
 Matching molecular weight (394.4 vs parent mass 394.

Added metadata based on: 
 Matching molecular weight (218.2 vs parent mass 218.1).Matching compound name ( Ser-Ile).
Added metadata based on: 
 Matching molecular weight (301.3 vs parent mass 301.1).Matching compound name ( Fenoxycarb).
Added metadata based on: 
 Matching molecular weight (336.4 vs parent mass 336.2).Matching compound name ( Acebutolol).
Added metadata based on: 
 Matching molecular weight (408.6 vs parent mass 408.3).Matching compound name ( Cholic acid).
Added metadata based on: 
 Matching molecular weight (360.4 vs parent mass 360.2).Matching compound name ( Prednisolone).
Added metadata based on: 
 Matching molecular weight (390.6 vs parent mass 390.3).Matching compound name ( Dioctyl phthalate).
Added metadata based on: 
 Matching molecular weight (292.3 vs parent mass 292.1).Matching compound name ( PyroGlu-Tyr).
Added metadata based on: 
 Matching molecular weight (217.2 vs parent mass 217.1).Matching compound name ( Pymetrozin).
Added metadata based on: 
 Match

Added metadata based on: 
 Matching molecular weight (704.0 vs parent mass 703.5).Matching compound name ( 1-Hexadecyl-2-(9Z-octadecenoyl)-sn-glycero-3-phosphoethanolamine).
Added metadata based on: 
 Matching molecular weight (786.1 vs parent mass 785.6).Matching compound name ( 1-Stearoyl-2-linoleoyl-sn-glycero-3-phosphocholine).
Added metadata based on: 
 Matching molecular weight (122.1 vs parent mass 122.0).Matching compound name ( Niacinamide).
Added metadata based on: 
 Matching molecular weight (222.2 vs parent mass 222.1).Matching compound name ( Fraxidin).
Added metadata based on: 
 Matching molecular weight (351.4 vs parent mass 351.1).Matching compound name ( Benflurex).
Added metadata based on: 
 Matching molecular weight (204.2 vs parent mass 204.1).Matching compound name ( L-Tryptophan).
Added metadata based on: 
 Matching molecular weight (339.4 vs parent mass 339.1).Matching compound name ( Papaverine).
Added metadata based on: 
 Matching molecular weight (274.4 vs par

Added metadata based on: 
 Matching molecular weight (458.4 vs parent mass 458.1).Matching compound name ( Epigallocatechin gallate).
Added metadata based on: 
 Matching molecular weight (255.4 vs parent mass 255.3).Matching compound name ( Palmitamide).
Added metadata based on: 
 Matching molecular weight (188.2 vs parent mass 188.1).Matching compound name ( Gly-Leu).
Added metadata based on: 
 Matching molecular weight (230.3 vs parent mass 230.1).Matching compound name ( Kavain, DL-).
Added metadata based on: 
 Matching molecular weight (360.8 vs parent mass 360.1).Matching compound name ( Fenofibrate).
Added metadata based on: 
 Matching molecular weight (523.7 vs parent mass 523.4).Matching compound name ( 1-Octadecanoyl-sn-glycero-3-phosphocholine).
Added metadata based on: 
 Matching molecular weight (178.2 vs parent mass 178.1).Matching compound name ( Coniferyl aldehyde).
Added metadata based on: 
 Matching molecular weight (546.8 vs parent mass 546.4).Matching compound name (

Added metadata based on: 
 Matching molecular weight (744.0 vs parent mass 743.5).Matching compound name ( 1-Stearoyl-2-linoleoyl-sn-glycero-3-phosphoethanolamine).
Added metadata based on: 
 Matching molecular weight (414.5 vs parent mass 414.2).Matching compound name ( Nonaethylene glycol).
Added metadata based on: 
 Matching molecular weight (273.3 vs parent mass 273.2).Matching compound name ( Arg-Val).
Added metadata based on: 
 Matching molecular weight (390.6 vs parent mass 390.3).Matching compound name ( Dioctyl phthalate).
Added metadata based on: 
 Matching molecular weight (244.3 vs parent mass 244.1).Matching compound name ( Biotin).
Added metadata based on: 
 Matching molecular weight (252.3 vs parent mass 252.1).Matching compound name ( His-Pro).
Added metadata based on: 
 Matching molecular weight (188.3 vs parent mass 188.1).Matching compound name ( 3-Hydroxycapric acid).
Added metadata based on: 
 Matching molecular weight (278.4 vs parent mass 278.2).Matching compound

Added metadata based on: 
 Matching molecular weight (198.2 vs parent mass 198.1).Matching compound name ( Syringic acid).
Added metadata based on: 
 Matching molecular weight (775.0 vs parent mass 774.5).Matching compound name ( Nis(monooleoylglycero)phosphate (S,R Isomer)).
Added metadata based on: 
 Matching molecular weight (310.3 vs parent mass 310.1).Matching compound name ( Sulfadimethoxine).
Added metadata based on: 
 Matching molecular weight (239.3 vs parent mass 239.2).Matching compound name ( Albuterol).
Added metadata based on: 
 Matching molecular weight (515.7 vs parent mass 515.3).Matching compound name ( Taurocholic acid).
Added metadata based on: 
 Matching molecular weight (284.7 vs parent mass 284.0).Matching compound name ( Sulfachloropyridazine).
Added metadata based on: 
 Matching molecular weight (268.4 vs parent mass 268.2).Matching compound name ( cis-7-Hexadecenoic acid methyl ester).
Added metadata based on: 
 Matching molecular weight (181.2 vs parent mass 

Added metadata based on: 
 Matching molecular weight (304.5 vs parent mass 304.2).Matching compound name ( cis-5,8,11,14-Eicosatetraenoic acid).
Added metadata based on: 
 Matching molecular weight (163.2 vs parent mass 163.1).Matching compound name ( 1H-Purin-6-amine, N,N-dimethyl-).
Added metadata based on: 
 Matching molecular weight (332.5 vs parent mass 332.3).Matching compound name ( Arachidonic acid ethyl ester).
Added metadata based on: 
 Matching molecular weight (290.3 vs parent mass 290.1).Matching compound name ( (-)-Epicatechin).
Added metadata based on: 
 Matching molecular weight (280.4 vs parent mass 280.2).Matching compound name ( Linoelaidic acid).
Added metadata based on: 
 Matching molecular weight (720.0 vs parent mass 719.6).Matching compound name ( 1,2-Dihexadecanoyl-sn-glycero-3-phospho-N,N-dimethylethanolamine).
Added metadata based on: 
 Matching molecular weight (205.2 vs parent mass 205.1).Matching compound name ( Dexpanthenol).
Added metadata based on: 
 Ma

Added metadata based on: 
 Matching molecular weight (122.1 vs parent mass 122.0).Matching compound name ( Niacinamide).
Added metadata based on: 
 Matching molecular weight (390.6 vs parent mass 390.3).Matching compound name ( Diisooctyl phthalate).
Added metadata based on: 
 Matching molecular weight (356.5 vs parent mass 356.3).Matching compound name ( Monoolein).
Added metadata based on: 
 Matching molecular weight (197.2 vs parent mass 197.1).Matching compound name ( 3,4-Dihydroxy-L-phenylalanine).
Added metadata based on: 
 Matching molecular weight (228.2 vs parent mass 228.1).Matching compound name ( Resveratrol).
Added metadata based on: 
 Matching molecular weight (278.3 vs parent mass 278.2).Matching compound name ( Dibutyl phthalate).
Added metadata based on: 
 Matching molecular weight (302.5 vs parent mass 302.2).Matching compound name ( Isopimaric acid).
Added metadata based on: 
 Matching molecular weight (175.2 vs parent mass 175.1).Matching compound name ( L-Citrullin

Added metadata based on: 
 Matching molecular weight (734.0 vs parent mass 733.6).Matching compound name ( 1-Stearoyl-2-myristoyl-sn-glycero-3-phosphocholine).
Added metadata based on: 
 Matching molecular weight (230.2 vs parent mass 230.1).Matching compound name ( Asp-Pro).
Added metadata based on: 
 Matching molecular weight (792.1 vs parent mass 791.5).Matching compound name ( 1-Octadecanoyl-2-(4Z,7Z,10Z,13Z,16Z,19Z-docosahexaenoyl)-sn-glycero-3-phosphoethanolamine).
Added metadata based on: 
 Matching molecular weight (310.3 vs parent mass 310.1).Matching compound name ( .alpha.-L-Glu-L-Tyr).
Added metadata based on: 
 Matching molecular weight (352.8 vs parent mass 352.1).Matching compound name ( Griseofulvin).
Added metadata based on: 
 Matching molecular weight (329.2 vs parent mass 329.1).Matching compound name ( Adenosine 3',5'-cyclic monophosphate).
Added metadata based on: 
 Matching molecular weight (187.3 vs parent mass 187.2).Matching compound name ( N8-Acetylspermidine)

In [9]:
# Check outcome of PubChem lookups
print("--- first 20,000 spectra")
print("--- before pubchem lookups:")
count_annotations(reference_spectrums[:20000])
print("--- after pubchem lookups:")
count_annotations(reference_spectrums_lookup)

--- first 20,000 spectra
--- before pubchem lookups:
Inchis: 15764 -- 9212 unique
Smiles: 15784 -- 9802 unique
Inchikeys: 15769 -- 8658 unique
--- after pubchem lookups:
Inchis: 17027 -- 9756 unique
Smiles: 17034 -- 10502 unique
Inchikeys: 17027 -- 9131 unique


### The above lookup process takes a night to run, so here importing the final results:

In [15]:
# Load fully processed dataset from elsewhere
from matchms.importing.load_from_json import load_from_json
reference_spectrums_lookup = load_from_json(os.path.join(path_data,'gnps_all_after_pubchem_lookups.json'))

In [16]:
count_annotations(reference_spectrums_lookup)

Inchis: 92135 -- 15813 unique
Smiles: 92127 -- 20461 unique
Inchikeys: 92135 -- 13688 unique


### Closer look at example that was changed based on name + parent mass

In [12]:
for i, spec in enumerate(reference_spectrums_lookup[2000:3000]):
    log_entry = spec.get("metadata_added_based_on")
    if log_entry:
        print(i)
        break

614


### before pubchem lookup

In [17]:
reference_spectrums[2614].metadata

{'pepmass': [545.0, None],
 'charge': 1,
 'mslevel': '2',
 'source_instrument': 'LC-ESI-Ger',
 'filename': '20111105_Jenia_Ger_Scytonemin_MH_545_qb.1.1..mgf',
 'seq': '*..*',
 'notes': '',
 'ionmode': 'positive',
 'organism': 'GNPS-LIBRARY',
 'name': 'Scytonemin M+H',
 'smiles': '',
 'inchi': '',
 'inchiaux': 'N/A',
 'libraryquality': '1',
 'spectrumid': 'CCMSLIB00000001550',
 'activation': 'CID',
 'instrument': 'ion trap',
 'title': 'Scan Number: 1',
 'scans': '1',
 'adduct': 'M+H',
 'parent_mass': 543.9927235480093,
 'inchikey': ''}

### After pubchem lookup
--> New metadata added and log entry:  
``'metadata_added_based_on': 'PubChem match: Matching molecular weight (544.6 vs parent mass 544.0).Matching compound name (Scytonemin).'`` 

In [18]:
reference_spectrums_lookup[2614].metadata

{'pepmass': [545.0, None],
 'charge': 1,
 'mslevel': '2',
 'source_instrument': 'LC-ESI-Ger',
 'filename': '20111105_Jenia_Ger_Scytonemin_MH_545_qb.1.1..mgf',
 'seq': '*..*',
 'notes': '',
 'ionmode': 'positive',
 'organism': 'GNPS-LIBRARY',
 'name': 'Scytonemin M+H',
 'smiles': 'C1=CC=C2N=C\\3C(=C(C(=O)/C3=C/C4=CC=C(C=C4)O)C5=C6C(=NC7=CC=CC=C67)/C(=C\\C8=CC=C(C=C8)O)/C5=O)C2=C1',
 'inchi': 'InChI=1S/C36H20N2O4/c39-21-13-9-19(10-14-21)17-25-33-29(23-5-1-3-7-27(23)37-33)31(35(25)41)32-30-24-6-2-4-8-28(24)38-34(30)26(36(32)42)18-20-11-15-22(40)16-12-20/h1-18,39-40H/b25-17+,26-18+',
 'inchiaux': 'N/A',
 'libraryquality': '1',
 'spectrumid': 'CCMSLIB00000001550',
 'activation': 'CID',
 'instrument': 'ion trap',
 'title': 'Scan Number: 1',
 'scans': '1',
 'adduct': 'M+H',
 'parent_mass': 543.9927235480093,
 'inchikey': 'CGZKSPLDUIRCIO-RPCRKUJJSA-N',
 'metadata_added_based_on': 'PubChem match: Matching molecular weight (544.6 vs parent mass 544.0).Matching compound name (Scytonemin).'}

### Select positive mode spectra only

In [8]:
reference_spectrums_positive = [s.clone() for s in reference_spectrums if s.metadata["ionmode"] == 'positive']